packages/react/src/components/Accordion/AccordionItem.js

```diff
@@ -38,16 +40,27 @@ function AccordionItem({
     setPrevIsOpen(open);
   }
 
-  // When the AccordionItem heading is clicked, toggle the open state of the
-  // panel
-  function onClick(event) {
-    const nextValue = !isOpen;
-    setIsOpen(nextValue);
-    if (onHeadingClick) {
+  useEffect(() => {
+    savedOnHeadingClick.current = onHeadingClick;
+  }, [onHeadingClick]);
+
+  useEffect(() => {
```

Comment:
Looks that this is a use case of `useCallback()`, where callback function can be updated upon change in closure refs.

Reply:
How would you suggest the code be structured instead?

If I understand correctly, `useCallback` is often used by the owner for passing a prop to a component. In other words, I would pass a closure using `useCallback` to a component versus the component using `useCallback`.

In this case, the ref pattern for saving the callback is coming from: https://overreacted.io/making-setinterval-declarative-with-react-hooks/#just-show-me-the-code 

```js
function useInterval(callback, delay) {
  const savedCallback = useRef();

  // Remember the latest callback.
  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);
```

Reply:
The link seems to use `useRef()` to remember the latest ref of `callback` to ensure `setInterval()` always runs the latest callback, given React doesn't have control over `setInterval()` to make it detect change in `callback`.

OTOH, the only changing factor of `onClick()` call is the set of React props, where React (hooks) have complete control over. That said, we can do:

```javascript
const onClick = useCallback(event => {
  const nextValue = !isOpen;
  setIsOpen(nextValue);
  if (onHeadingClick) {
    // TODO: normalize signature, potentially:
    // onHeadingClick :: (event: Event, state: { isOpen: Boolean }) => any
    onHeadingClick({ isOpen: nextValue, event });
  }
}, [isOpen, onHeadingClick]);

const onKeyDown = useCallback(event => {
  if (isOpen && match(event, keys.Escape)) {
    setIsOpen(false);
  }
}, [isOpen]);
```

Reply:
I think useCallback is unnecessary here:
> useCallback is useful when passing callbacks to optimized child components that rely on reference equality to prevent unnecessary renders (e.g. shouldComponentUpdate).
> useCallback(fn, deps) is equivalent to useMemo(() => fn, deps).

I think we should be careful to use the hooks for their intended purpose (effect for imperative side effects and callback for memoized callbacks.) using them to hack the lifecycle order sounds a whole lot like the gDSDP fiasco.

Reply:
The use I brought up above _is_ for the intended purpose. If you re-create event listener for every render then unnecessary render will happen by different event listener reference. `useCallback()` helps us prevent that by changing the event listener function only when the properties the event listener depends on changes.

Reply:
It seems like both of you are expressing that it should be used for the same purpose (preventing unnecessary re-renders) but maybe disagree on when to use it?

The underlying bug I was trying to address in this diff was actually not a bug on our end, which further caused confusion I think by how I attempted to remedy that 🤦‍♂️